x/stake/keeper/validator.go

```diff
@@ -199,78 +199,101 @@ func (k Keeper) ClearTendermintUpdates(ctx sdk.Context) {
+	if oldFound && oldValidator.PoolShares.Bonded().LT(newValidator.PoolShares.Bonded()) {
+		return true
 	}
+	return false
+}
+
+func (k Keeper) getBondDetails(ctx sdk.Context, oldFound bool, oldValidator,
+	newValidator types.Validator) (height int64, intraTxCounter int16) {
 
 	// if already a validator, copy the old block height and counter, else set them
 	if oldFound && oldValidator.Status() == sdk.Bonded {
-		validator.BondHeight = oldValidator.BondHeight
-		validator.BondIntraTxCounter = oldValidator.BondIntraTxCounter
-	} else {
-		validator.BondHeight = ctx.BlockHeight()
-		counter := k.GetIntraTxCounter(ctx)
-		validator.BondIntraTxCounter = counter
-		k.SetIntraTxCounter(ctx, counter+1)
+		height = oldValidator.BondHeight
+		intraTxCounter = oldValidator.BondIntraTxCounter
+		return
 	}
+	height = ctx.BlockHeight()
+	counter := k.GetIntraTxCounter(ctx)
+	intraTxCounter = counter
+	k.SetIntraTxCounter(ctx, counter+1)
+	return
+}
+
+func (k Keeper) updateValidatorPower(ctx sdk.Context, oldFound bool, oldValidator,
+	newValidator types.Validator, pool types.Pool) (valPower []byte) {
+	store := ctx.KVStore(k.storeKey)
 
 	// update the list ordered by voting power
 	if oldFound {
 		store.Delete(GetValidatorsByPowerIndexKey(oldValidator, pool))
 	}
-	valPower := GetValidatorsByPowerIndexKey(validator, pool)
-	store.Set(valPower, validator.Owner)
-
-	// efficiency case:
-	// if already bonded and power increasing only need to update tendermint
-	if powerIncreasing && !validator.Revoked && oldValidator.Status() == sdk.Bonded {
-		bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
-		store.Set(GetTendermintUpdatesKey(ownerAddr), bz)
-		return validator
-	}
+	valPower = GetValidatorsByPowerIndexKey(newValidator, pool)
+	store.Set(valPower, newValidator.Owner)
 
-	// efficiency case:
-	// if was unbonded/or is a new validator - and the new power is less than the cliff validator
-	cliffPower := k.GetCliffValidatorPower(ctx)
-	if cliffPower != nil &&
-		(!oldFound || (oldFound && oldValidator.Status() == sdk.Unbonded)) &&
-		bytes.Compare(valPower, cliffPower) == -1 { //(valPower < cliffPower
-		return validator
-	}
-
-	// update the validator set for this validator
-	updatedVal := k.UpdateBondedValidators(ctx, validator)
-	if updatedVal.Owner != nil { // updates to validator occurred  to be updated
-		validator = updatedVal
-	}
-	// if decreased in power but still bonded, update Tendermint validator
-	// (if updatedVal is set, the validator has changed bonding status)
-	stillBonded := oldFound && oldValidator.Status() == sdk.Bonded && updatedVal.Owner == nil
-	if stillBonded && oldValidator.PoolShares.Bonded().GT(validator.PoolShares.Bonded()) {
-		bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
-		store.Set(GetTendermintUpdatesKey(ownerAddr), bz)
-	}
-	return validator
+	return valPower
 }
 
 // Update the validator group and kick out any old validators. In addition this
```

Comment:
Does this help make things a bit more clear?
```
// UpdateBondedValidators updates the validator group and kicks out any old
// validators. In addition, a validator may be added which has updated it's
// bonded tokens to the validator group. This validator is specified through
// the updatedValidatorAddr term.
```

Reply:
going to leave it, doesn't really improve IMO - but this comment should be updated more

Reply:
updated this comment to read:

```golang 
// Update the bonded validator group based on a change to the validator
// affectedValidator. This function potentially adds the affectedValidator to
// the bonded validator group which kicks out the cliff validator. Under this
// situation this function returns the updated affectedValidator.
//
// The correct bonded subset of validators is retrieved by iterating through an
// index of the validators sorted by power, stored using the
// ValidatorsByPowerIndexKey.  Simultaneously the current validator records are
// updated in store with the ValidatorsBondedIndexKey. This store is used to
// determine if a validator is a validator without needing to iterate over all
// validators.

```